Fixes racy oximeter integration test

bnaecker opened the pull request:
Fixes a race in the oximeter integration test that verifies behavior
when the producer or collector dies and comes back. The race can occur
when the collector scrapes a sample, then the producer is dropped. At
this point, the test gets the new timeseries, and asserts it's the same
as the old. That's not quite right, since it's possible for the new
sample to have been produced immediately before the producer is dropped.
Then it appears on the next query, and the assertion is tripped.

The test is modified to be slightly weaker. It asserts that the
timeseries match up to the number of samples previously collected, and
that any new samples have timestamps before the time the producer is
dropped.

ahl commented on nexus/tests/integration_tests/oximeter.rs:
```diff
@@ -146,12 +146,52 @@ async fn test_oximeter_reregistration() {
+    //
+    // Specifically, we grab a timestamp before dropping the producer, and assert that any data in
+    // the timeseries is from before that timestamp.
+    let time_producer_dropped = chrono::Utc::now();
```
is it worth noting that a very narrow race still exists insofar as data might be recorded between this line and the actual `drop`? ... if I understand that correctly that is...

ahl commented on nexus/tests/integration_tests/oximeter.rs:
```diff
@@ -146,12 +146,52 @@ async fn test_oximeter_reregistration() {
+    // the collector's memory. So we first check that the original timeseries is a prefix of the
+    // new timeseries. Then we check that any remaining measurements are no later than the
+    // timestamp immediately prior to dropping the producer.
+    fn check_following_timeseries(
```
Might you want #[track_caller] on this function so that assertion failures would indicate which of the locations was problematic? or would it be better to just enable stacks at that point

ahl reviewed (commented):
Thanks for fixing this!

davepacheco commented:
Can one of you paste the output from a failed test?  I wanted to see if an existing test failure ([this one](https://github.com/oxidecomputer/omicron/runs/4958955351?check_suite_focus=true#step:11:957)) is the same one.

davepacheco commented:
I see that from the changes here that it probably is the same test.  It'd still be nice to have output here in case someone else sees the failure and searches issues for it.

bnaecker commented:
@davepacheco That's a different failure, for better or worse. The output that Adam saw is below:

```
---- integration_tests::oximeter::test_oximeter_reregistration stdout ----
log file: "/var/folders/q4/wlt470150sn72ty0zyxxcng8003c85/T/test_all-f0e4e73ee023ac59-test_oximeter_reregistration.10024.30.log"
note: configured to log to "/var/folders/q4/wlt470150sn72ty0zyxxcng8003c85/T/test_all-f0e4e73ee023ac59-test_oximeter_reregistration.10024.30.log"
thread 'integration_tests::oximeter::test_oximeter_reregistration' panicked at 'assertion failed: `(left == right)`
  left: `Timeseries { timeseries_name: "integration_target:integration_metric", target: Target { name: "integration_target", fields: [Field { name: "name", value: String("integration-test-target") }, Field { name: "name", value: String("integration-test-target") }] }, metric: Metric { name: "integration_metric", fields: [], datum_type: I64 }, measurements: [Measurement { timestamp: 2022-01-26T00:31:13.463940Z, datum: I64(0) }, Measurement { timestamp: 2022-01-26T00:31:14.462231Z, datum: I64(1) }, Measurement { timestamp: 2022-01-26T00:31:15.462622Z, datum: I64(2) }, Measurement { timestamp: 2022-01-26T00:31:16.461730Z, datum: I64(3) }] }`,
 right: `Timeseries { timeseries_name: "integration_target:integration_metric", target: Target { name: "integration_target", fields: [Field { name: "name", value: String("integration-test-target") }, Field { name: "name", value: String("integration-test-target") }] }, metric: Metric { name: "integration_metric", fields: [], datum_type: I64 }, measurements: [Measurement { timestamp: 2022-01-26T00:31:13.463940Z, datum: I64(0) }, Measurement { timestamp: 2022-01-26T00:31:14.462231Z, datum: I64(1) }, Measurement { timestamp: 2022-01-26T00:31:15.462622Z, datum: I64(2) }, Measurement { timestamp: 2022-01-26T00:31:16.461730Z, datum: I64(3) }, Measurement { timestamp: 2022-01-26T00:31:17.462050Z, datum: I64(4) }, Measurement { timestamp: 2022-01-26T00:31:18.463550Z, datum: I64(5) }] }`', nexus/tests/integration_tests/oximeter.rs:254:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
WARN: dropped CockroachInstance without cleaning it up first (there may still be a child process running and a temporary directory leaked)
WARN: temporary directory leaked: /var/folders/q4/wlt470150sn72ty0zyxxcng8003c85/T/.tmpzpsVbQ
WARN: dropped ClickHouseInstance without cleaning it up first (there may still be a child process running and a temporary directory leaked)

```
